### PR TITLE
rgw: implement stringlike wildcard for bucket policy

### PR DESCRIPTION
**Pull request by joke-lee**
hi, @cbodley ,@adamemerson recently i implement stringlike wildcard for bucket policy,  do you mind to help review this ? thanks

### CODE REVIEW

**Comment by joke-lee** [comment]
implement http://tracker.ceph.com/issues/20308

File: src/rgw/rgw_iam_policy.h [hunk 1]
```diff
@@ -359,6 +360,18 @@ struct Condition {
     }
   };
 
+  struct ci_string_like : public std::binary_function<const std::string,
+                                                     const std::string,
+                                                     bool> {
+    bool operator ()(const std::string& s1,
+                    const std::string& s2) const {
+      if(fnmatch(s2.c_str(), s1.c_str(), 0) == 0){
```
Comment (cbodley) [inline src/rgw/rgw_iam_policy.h]: it doesn't look like [fnmatch()](http://man7.org/linux/man-pages/man3/fnmatch.3.html) defaults to case-insensitive matching:
> FNM_CASEFOLD: If this flag (a GNU extension) is set, the pattern is matched case-insensitively.

**Comment by joke-lee** [comment]
test this please

**Comment by cbodley** [comment]
make check failures should be fixed with #15730, can you rebase please?

File: src/rgw/rgw_iam_policy.h [hunk 2]
```diff
@@ -359,6 +360,18 @@ struct Condition {
     }
   };
 
+  struct ci_string_like : public std::binary_function<const std::string,
+                                                     const std::string,
+                                                     bool> {
+    bool operator ()(const std::string& s1,
+                    const std::string& s2) const {
+      if(fnmatch(s2.c_str(), s1.c_str(), FNM_CASEFOLD) == 0){
+        return true;
```
Comment (adamemerson) [inline src/rgw/rgw_iam_policy.h]: It may be better to use the globbing function from rgw_common so we can get case sensitivity as appropriate.

Comment (joke-lee) [inline src/rgw/rgw_iam_policy.h]: hi @adamemerson , i test on amazon s3. 
the test policy set as follow
```
{
  "Id": "Policy1497346096739",
  "Version": "2012-10-17",
  "Statement": [
    {
      "Sid": "Stmt1497346083540",
      "Action": [
        "s3:GetObject"
      ],
      "Effect": "Allow",
      "Resource": "arn:aws:s3:::yly-policy/*",
      "Condition": {
        "StringLike": {
          "aws:Referer": "http://WWW.example.com/*"
        }
      },
      "Principal": "*"
    }
  ]
}
```
and curl with referer `http://www.example.com/`

```
> GET /yly-policy/1.txt HTTP/1.1
> User-Agent: curl/7.29.0
> Host: s3.amazonaws.com
> Accept: */*
> referer: http://www.example.com/
> 
< HTTP/1.1 403 Forbidden
< x-amz-request-id: FC9CE4656FA773F7
< x-amz-id-2: 2j6il8iRZowlAvP4ylLP2XFGS2ZAIpdZ9Mfb6JnQxiTts1SHnw6fEU6bk+jDmd0Rq6yK/3zBR98=
< Content-Type: application/xml
< Transfer-Encoding: chunked
< Date: Sat, 17 Jun 2017 13:02:09 GMT
< Server: AmazonS3

```


and curl with refer `http://WWW.example.com/`
```
> GET /yly-policy/1.txt HTTP/1.1
> User-Agent: curl/7.29.0
> Host: s3.amazonaws.com
> Accept: */*
> referer: http://WWW.example.com/
> 
< HTTP/1.1 200 OK
< x-amz-id-2: iuOA7wDe/OQoukdM+2MepD9g99GuP0Vkm2raDIYNSZSTt6M3AnDI+ztNX/5yDrr1+Tjt0cT1t/M=
< x-amz-request-id: AF2E4CA0DA7C49B2
< Date: Sat, 17 Jun 2017 13:02:48 GMT
< Last-Modified: Tue, 13 Jun 2017 08:43:37 GMT
< ETag: "70e2c79c9f37189d91b784edb36077ce"
< Accept-Ranges: bytes
< Content-Type: binary/octet-stream
< Content-Length: 273
< Server: AmazonS3

```

so i wonder why we need  get case sensitivity ?

File: src/rgw/rgw_iam_policy.h (hunk 1 above)
Comment (joke-lee) [inline src/rgw/rgw_iam_policy.h]: hi @cbodley  i change the function ci_string_like to string_like ,  i think  string_like should be case-sensitively

**Comment by joke-lee** [comment]
hi @adamemerson ,if we use the match function in rgw_common
and we set reffer string like policy.
http://www.example.com/  could not match pattern http://www.example.com/* 
so i think fnmatch is better.

**Comment by adamemerson** [comment]
Just add the MATCH_POLICY_ACTION flag to it and it will ignore case. (This should probably be renamed sometime.)

**Comment by mattbenjamin** [comment]
@joke-lee will add MATCH_POLICY_ACTION update?

Matt

**Comment by cbodley** [comment]
what are we blocked on  here? we'll need wildcard support to pass https://github.com/ceph/s3-tests/pull/167:
```
                    "StringLikeIfExists": {
                        "aws:Referer": "http://www.example.com/*"
                    }
```

**Comment by joke-lee** [comment]
@cbodley compile error if we use match funtion  and  MATCH_POLICY_ACTION defined in rgw_common.h,i still look into this.

**Comment by cbodley** [comment]
hi @joke-lee, i was able to get this working in https://github.com/ceph/ceph/pull/16491, could you take a look? i also opened https://github.com/ceph/ceph/issues/16463 with some extra fixes for `IfExists` to get your s3test in https://github.com/ceph/s3-tests/pull/167 to pass

**Pull request closed without merging**